### PR TITLE
[ZEPPELIN-2116]. livy interpreter needs to restart after previous session timeout in secured cluster

### PR DESCRIPTION
**Pull request by zjffdu**
What is this PR for?

It throw different exception in secured cluster. This PR fix the issue in secured cluster

What type of PR is it?

[Bug Fix]

Todos

 - Task
What is the Jira issue?

https://issues.apache.org/jira/browse/ZEPPELIN-2116
How should this be tested?

Tested manually.

Questions:

Does the licenses files need update? No
Is there breaking changes for older versions? No
Does this needs documentation? no

### CODE REVIEW

**Comment by zjffdu** [comment]
@prabhjyotsingh @felixcheung Please help review, CI failure is not relevant. 

```
TypeError: namedtuple() missing 3 required keyword-only arguments: 'verbose', 'rename', and 'module'
, %text pyspark is not responding] expected:<SUCCESS> but was:<ERROR>
  PySparkInterpreterMatplotlibTest.showPlot:161 [%text pyspark is not responding] expected:<SUCCESS> but was:<ERROR>
  PySparkInterpreterMatplotlibTest.testAngular:230 [%text pyspark is not responding] expected:<SUCCESS> but was:<ERROR>
  PySparkInterpreterMatplotlibTest.testClose:184 expected:<0> but was:<1>
  PySparkInterpreterMatplotlibTest.testNoClose:211 [%text pyspark is not responding] expected:<SUCCESS> but was:<ERROR>
  PySparkInterpreterTest.testBasicIntp:111 expected:<SUCCESS> but was:<ERROR>
  PySparkInterpreterTest.testCompletion:120 null
Tests run: 7, Failures: 7, Errors: 0, Skipped: 0
```

File: livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java [hunk 1]
```diff
@@ -407,6 +408,13 @@ private String callRestAPI(String targetURL, String method, String jsonData)
       response = new ResponseEntity(e.getResponseBodyAsString(), e.getStatusCode());
       LOGGER.error(String.format("Error with %s StatusCode: %s",
           response.getStatusCode().value(), e.getResponseBodyAsString()));
+    } catch (RestClientException e) {
+      // Exception happens when kerberos is enabled.
+      if (e.getMessage().contains("404 Not Found")) {
+        throw new SessionNotFoundException("Session not found when kerberos is enabled ");
```
Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: how is `RestClientException` map to `kerberos`?

Comment (zjffdu) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: When kerberos is enabled,  KerberosRestTemplate will throw RestClientException while RestTemplate throw HttpClientErrorException when kerberos is disabled. You can check the jira description which has the stracktrace.

Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: right, it has the stack but not the exception type.
it just seems somewhat generic - I assume RestClientException could be possible in other error cases too? Is there an inner exception that we could use to scope this down?

Comment (zjffdu) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: Oops, seem the stacktrace in jira description is not complete. Here's the full stacktrace

```
ERROR [2017-02-16 07:30:03,220] ({pool-2-thread-5} Job.java[run]:188) - Job failed
org.springframework.web.client.RestClientException: Error running rest call; nested exception is org.springframework.web.client.HttpClientErrorException: 404 Not Found
	at org.springframework.security.kerberos.client.KerberosRestTemplate.doExecute(KerberosRestTemplate.java:196)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:580)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:498)
	at org.apache.zeppelin.livy.BaseLivyInterprereter.callRestAPI(BaseLivyInterprereter.java:419)
	at org.apache.zeppelin.livy.BaseLivyInterprereter.executeStatement(BaseLivyInterprereter.java:380)
	at org.apache.zeppelin.livy.BaseLivyInterprereter.interpret(BaseLivyInterprereter.java:250)
	at org.apache.zeppelin.livy.BaseLivyInterprereter.interpret(BaseLivyInterprereter.java:144)
	at org.apache.zeppelin.interpreter.LazyOpenInterpreter.interpret(LazyOpenInterpreter.java:94)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer$InterpretJob.jobRun(RemoteInterpreterServer.java:489)
	at org.apache.zeppelin.scheduler.Job.run(Job.java:175)
	at org.apache.zeppelin.scheduler.FIFOScheduler$1.run(FIFOScheduler.java:139)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.springframework.web.client.HttpClientErrorException: 404 Not Found
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:91)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:667)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:620)
	at org.springframework.security.kerberos.client.KerberosRestTemplate.doExecuteSubject(KerberosRestTemplate.java:202)
	at org.springframework.security.kerberos.client.KerberosRestTemplate.access$100(KerberosRestTemplate.java:67)
	at org.springframework.security.kerberos.client.KerberosRestTemplate$1.run(KerberosRestTemplate.java:191)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at org.springframework.security.kerberos.client.KerberosRestTemplate.doExecute(KerberosRestTemplate.java:187)
	... 17 more
```

Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: right, the stack is specific, but the exception `org.springframework.web.client.HttpClientErrorException: 404 Not Found` is very generic..

Comment (zjffdu) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: Great point, looks like the useful message is hidden in `HttpClientErrorException.getResponseBodyAsString()`, update the PR

**Comment by felixcheung** [comment]
LGTM
Waiting for CI and 
merging if no more comment

File: livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java [hunk 2]
```diff
@@ -46,6 +47,7 @@
 
   protected static final Logger LOGGER = LoggerFactory.getLogger(BaseLivyInterprereter.class);
   private static Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+  private static String SessionNotFound_Pattern = "\"Session '\\d+' not found.\"";
```
Comment (prabhjyotsingh) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java]: Camel case and underscore together ? Can you user all caps separated by underscore, like we are doing it in rest of the project for "static String" ?

**Comment by prabhjyotsingh** [comment]
Ran this couple of time on my local, works a expected.

**Comment by felixcheung** [comment]
I think that's a fair ask - @zjffdu could you update and we will merge?

**Comment by zjffdu** [comment]
@prabhjyotsingh @felixcheung  Thanks for review, the comments are addressed.

**Comment by felixcheung** [comment]
CI was passing here.
merged to master and branch-0.7

**Pull request closed without merging**